engine/crates/composition/src/compose/input_object.rs

```diff
@@ -7,53 +7,63 @@ pub(super) fn merge_input_object_definitions(
 ) {
     let description = definitions.iter().find_map(|def| def.description());
 
+    let composed_directives = collect_composed_directives(definitions.iter().map(|def| def.directives()), ctx);
+
+    ctx.insert_input_object(first.name(), description, composed_directives);
+
     // We want to take the intersection of the field sets.
-    let mut common_fields: Vec<StringId> = first.fields().map(|field| field.name().id).collect();
-    let mut fields_buf = HashSet::<StringId>::new();
+    let intersection: HashSet<StringId> = first
+        .fields()
+        .map(|field| field.name().id)
+        .filter(|field_name| definitions[1..].iter().all(|def| def.find_field(*field_name).is_some()))
+        .collect();
 
-    for input_object in definitions {
-        fields_buf.clear();
-        fields_buf.extend(input_object.fields().map(|f| f.name().id));
-        common_fields.retain(|field_name| fields_buf.contains(field_name));
-    }
+    let mut all_fields: Vec<_> = definitions
+        .iter()
+        .flat_map(|def| def.fields())
+        .map(|field| (field.name().id, field))
+        .collect();
 
-    // Check that no required field was excluded.
-    for field in definitions.iter().flat_map(|input_object| input_object.fields()) {
-        if field.r#type().is_required() && !common_fields.contains(&field.name().id) {
-            ctx.diagnostics.push_fatal(format!(
-                "The {input_type_name}.{field_name} field is not defined in all subgraphs, but it is required in {bad_subgraph}",
-                input_type_name = first.name().as_str(),
-                field_name = field.name().as_str(),
-                bad_subgraph = field.parent_definition().subgraph().name().as_str(),
-            ));
-        }
-    }
+    all_fields.sort_by_key(|(name, _)| *name);
 
-    let composed_directives = collect_composed_directives(definitions.iter().map(|def| def.directives()), ctx);
+    let mut start = 0;
 
-    ctx.insert_input_object(first.name(), description, composed_directives);
+    while start < all_fields.len() {
```

Comment:
I want to move all these blocks into separate functions, but let's do that in a separate PR to avoid noise in this one.